src/mpid/ch4/shm/posix/posix_win.h

```diff
@@ -409,27 +409,31 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_POSIX_mpi_win_allocate_hook(MPIR_Win * win)
     int mpi_errno = MPI_SUCCESS;
     MPIDI_POSIX_win_t *posix_win ATTRIBUTE((unused)) = NULL;
     MPIR_Comm *shm_comm_ptr = win->comm_ptr->node_comm;
+    bool mapfail_flag = false;
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_POSIX_MPI_WIN_ALLOCATE_HOOK);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_POSIX_MPI_WIN_ALLOCATE_HOOK);
 
     posix_win = &win->dev.shm.posix;
     posix_win->shm_mutex_ptr = NULL;
 
-    /* Enable shm RMA only when interprocess mutex is supported and
-     * more than 1 processes exist on the node. */
-    if (shm_comm_ptr == NULL || !MPL_proc_mutex_enabled())
+    /* Enable shm RMA only when interprocess mutex is supported,
+     * more than 1 processes exist on the node, and shm memory has been successfully allocated. */
+    if (shm_comm_ptr == NULL || !MPL_proc_mutex_enabled() || !MPIDI_CH4U_WIN(win, mmap_addr))
```

Comment:
This is a bit off topic but I think this line in MPL configure
https://github.com/pmodels/mpich/blob/master/src/mpl/configure.ac#L795
should be
```
if test "$ac_cv_func_pthread_mutexattr_setpshared" = "yes" ; then
```
instead of "no". This would affect the behavior of `MPL_proc_mutex_enabled()`, so if we are to fix it, I think it's worth including it in this PR and test it together.

Reply:
Also, should this change be really included in this patch?

Reply:
I think it makes sense to have the change in `posix_win` in this commit, as it is a part of the fallback routine when shm fails at win allocate (i.e., it disables load/store based RMA when shm allocation fails).

Let me fix the `ac_cv_func_pthread_mutexattr_setpshared` issue in a separate commit within this PR.

Reply:
After check again, I actually found that the following manual macro definition is unneeded. `AC_CHECK_FUNCS` will define it automatically.
```
    if test "$ac_cv_func_pthread_mutexattr_setpshared" = "no" ; then
      AC_DEFINE(HAVE_PTHREAD_MUTEXATTR_SETPSHARED,1, [Define if pthread_mutexattr_setpshared is available.])
    fi
```